File: apps/arc_identity/lib/arc/identity/sealed_box.ex
defmodule Arc.Identity.SealedBox do
  @moduledoc """
  Seal a message to a recipient's X25519 public key. No session is needed.
  Only the holder of the matching identity can open it.

  Format, version 1:

      <<1, ephemeral_pub::32, ciphertext_and_tag::binary>>

  Construction:

    1. Generate an ephemeral X25519 keypair.
    2. shared = ECDH(ephemeral_priv, recipient_pub)
    3. key = HKDF-SHA256(shared, salt: ephemeral_pub <> recipient_pub,
       info: "arc-sealed-v1", len: 32)
    4. ChaCha20-Poly1305 with a zero nonce and the ephemeral public key as
       associated data. The key is single use, so a fixed nonce is safe.

  `open/2` never reports why it failed. A wrong key and a corrupt body both
  return `{:error, :open_failed}`.
  """

  alias Arc.Identity
  alias Arc.Identity.HKDF

  @version 1
  @info "arc-sealed-v1"
  @nonce <<0::96>>
  @tag_bytes 16

  @doc """
  Seal `plaintext` to the recipient's X25519 public key.
  """
  @spec seal(<<_::256>>, binary()) :: binary()
  def seal(<<recipient_pub::binary-size(32)>>, plaintext) when is_binary(plaintext) do
    {ek_pub, ek_priv} = :crypto.generate_key(:ecdh, :x25519)
    shared = :crypto.compute_key(:ecdh, recipient_pub, ek_priv, :x25519)
    key = derive_key(shared, ek_pub, recipient_pub)

    {ciphertext, tag} =
      :crypto.crypto_one_time_aead(:chacha20_poly1305, key, @nonce, plaintext, ek_pub, true)

    <<@version, ek_pub::binary, ciphertext::binary, tag::binary>>
  end

  @doc """
  Open a sealed message with the recipient's identity.
  """
  @spec open(Identity.t(), binary()) :: {:ok, binary()} | {:error, :open_failed}
  def open(%Identity{} = identity, <<@version, ek_pub::binary-size(32), rest::binary>>)
      when byte_size(rest) >= @tag_bytes do
    {my_pub, my_priv} = Identity.to_x25519(identity)
    shared = :crypto.compute_key(:ecdh, ek_pub, my_priv, :x25519)
    key = derive_key(shared, ek_pub, my_pub)
    ct_size = byte_size(rest) - @tag_bytes
    <<ciphertext::binary-size(ct_size), tag::binary-size(@tag_bytes)>> = rest

    case :crypto.crypto_one_time_aead(
           :chacha20_poly1305,
           key,
           @nonce,
           ciphertext,
           ek_pub,
           tag,
           false
         ) do
      plaintext when is_binary(plaintext) -> {:ok, plaintext}
      :error -> {:error, :open_failed}
    end
  end

  def open(%Identity{}, sealed) when is_binary(sealed), do: {:error, :open_failed}

  defp derive_key(shared, ek_pub, recipient_pub) do
    HKDF.derive(shared, ek_pub <> recipient_pub, @info, 32)
  end
end

File: sealedbox/sealedbox.go
// Package sealedbox seals a message to a public key. No session is needed,
// and the recipient is offline while the sender works.
//
// Format, version 1:
//
//	<<1, ephemeral_public[32], ciphertext_and_tag>>
//
// Construction:
//
//  1. Generate an ephemeral X25519 keypair.
//  2. shared = ECDH(ephemeral_secret, recipient_public)
//  3. key = HKDF-SHA256(shared, salt: ephemeral_public || recipient_public,
//     info: "arc-sealed-v1", length: 32)
//  4. ChaCha20-Poly1305 with a zero nonce, and the ephemeral public key as
//     associated data. The key serves one message, so a fixed nonce is safe.
package sealedbox

import (
	"crypto/ecdh"
	"crypto/hkdf"
	"crypto/rand"
	"crypto/sha256"
	"errors"

	"github.com/gezibash/arc/identity"
	"golang.org/x/crypto/chacha20poly1305"
)

const (
	version  = 1
	info     = "arc-sealed-v1"
	keyBytes = 32
)

// ErrOpenFailed reports a sealed message that did not open. A wrong key and a
// changed body give the same error, so the answer tells an attacker nothing.
var ErrOpenFailed = errors.New("sealedbox: open failed")

var nonce = make([]byte, chacha20poly1305.NonceSize)

// Seal seals plaintext to the X25519 public key of a recipient.
func Seal(recipientPublic, plaintext []byte) ([]byte, error) {
	recipient, err := ecdh.X25519().NewPublicKey(recipientPublic)
	if err != nil {
		return nil, ErrOpenFailed
	}

	ephemeral, err := ecdh.X25519().GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}

	shared, err := ephemeral.ECDH(recipient)
	if err != nil {
		return nil, ErrOpenFailed
	}

	aead, err := newAEAD(shared, ephemeral.PublicKey().Bytes(), recipientPublic)
	if err != nil {
		return nil, err
	}

	out := make([]byte, 0, 1+identity.SeedBytes+len(plaintext)+aead.Overhead())
	out = append(out, version)
	out = append(out, ephemeral.PublicKey().Bytes()...)
	return aead.Seal(out, nonce, plaintext, ephemeral.PublicKey().Bytes()), nil
}

// SealTo seals plaintext to the Ed25519 public key of a recipient. The X25519
// key follows from that key, so a sender needs nothing else.
func SealTo(recipientEd25519, plaintext []byte) ([]byte, error) {
	recipient, err := identity.PublicKeyToX25519(recipientEd25519)
	if err != nil {
		return nil, err
	}
	return Seal(recipient, plaintext)
}

// Open opens a sealed message with the identity of the recipient.
func Open(id *identity.Identity, sealed []byte) ([]byte, error) {
	if len(sealed) < 1+identity.SeedBytes+chacha20poly1305.Overhead || sealed[0] != version {
		return nil, ErrOpenFailed
	}

	ephemeralPublic := sealed[1 : 1+identity.SeedBytes]
	body := sealed[1+identity.SeedBytes:]

	myPublic, mySecret, err := id.ToX25519()
	if err != nil {
		return nil, ErrOpenFailed
	}

	secret, err := ecdh.X25519().NewPrivateKey(mySecret)
	if err != nil {
		return nil, ErrOpenFailed
	}

	ephemeral, err := ecdh.X25519().NewPublicKey(ephemeralPublic)
	if err != nil {
		return nil, ErrOpenFailed
	}

	shared, err := secret.ECDH(ephemeral)
	if err != nil {
		return nil, ErrOpenFailed
	}

	aead, err := newAEAD(shared, ephemeralPublic, myPublic)
	if err != nil {
		return nil, ErrOpenFailed
	}

	plaintext, err := aead.Open(nil, nonce, body, ephemeralPublic)
	if err != nil {
		return nil, ErrOpenFailed
	}
	return plaintext, nil
}

func newAEAD(shared, ephemeralPublic, recipientPublic []byte) (interface {
	Seal(dst, nonce, plaintext, additionalData []byte) []byte
	Open(dst, nonce, ciphertext, additionalData []byte) ([]byte, error)
	Overhead() int
}, error) {
	salt := make([]byte, 0, len(ephemeralPublic)+len(recipientPublic))
	salt = append(salt, ephemeralPublic...)
	salt = append(salt, recipientPublic...)

	key, err := hkdf.Key(sha256.New, shared, salt, info, keyBytes)
	if err != nil {
		return nil, err
	}
	return chacha20poly1305.New(key)
}
